### DISCUSSION TITLE
SQLAlchemy 2.0: Typing support for hyrbrid property

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#9321] [Usage Questions] [answered]
Has typing support improved for the `hyrid_property`? In pre-2.0 we had to use the following workaround:
```python
if TYPE_CHECKING:
    typed_hybrid_property = property
else:
    from sqlalchemy.ext.hybrid import hybrid_property as typed_hybrid_property
```
so that mypy does not complain.

**SQLAlchemy 2.0.3**

**Comment by zzzeek** [comment, marked as answer]
yes, typing support has improved.  however, to keep mypy happy, you have to use a different name for your setter vs. getter.  and.....I am completely alarmed that we have not updated the hybrid docs for this.   wow, I just got finished telling someone we had migrated all our mappings in the docs.   let's fix that

**Reply by zzzeek** [reply]
Here is the lead example in the docs with annotations.  docs are ongoing and I also have adjusted a few types in hybrid.py so far to accommodate for some of the examples.   below should work right now


```py
from __future__ import annotations

from sqlalchemy import ColumnElement
from sqlalchemy import Float
from sqlalchemy import func
from sqlalchemy import type_coerce
from sqlalchemy.ext.hybrid import hybrid_method
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

class Base(DeclarativeBase):
    pass

class Interval(Base):
    __tablename__ = 'interval'

    id: Mapped[int] = mapped_column(primary_key=True)
    start: Mapped[int]
    end: Mapped[int]

    def __init__(self, start: int, end: int):
        self.start = start
        self.end = end

    @hybrid_property
    def length(self) -> int:
        return self.end - self.start

    @hybrid_method
    def contains(self, point: int) -> bool:
        return (self.start <= point) & (point <= self.end)

    @hybrid_method
    def intersects(self, other: Interval) -> bool:
        return self.contains(other.start) | self.contains(other.end)

    @hybrid_property
    def _radius_getter(self) -> float:
        return abs(self.length) / 2

    @_radius_getter.setter
    def _radius_setter(self, value: float) -> None:
        self.length = value * 2

    @_radius_setter.expression
    def radius(cls) -> ColumnElement[float]:
        return type_coerce(func.abs(cls.length) / 2, Float)

```

**Reply by michaelbukachi** [reply]
Awesome. Thanks!

**Reply by CaselIT** [reply]
> @_radius_setter.expression

shouldn't this usually be `@_radius_getter.expression`?

**Reply by zzzeek** [reply]
it all has to build on the previous one, _radius_getter became _radius_setter.

blame https://github.com/python/typing/discussions/1102 for this sad state of affairs

**Reply by CaselIT** [reply]
interesting, I would have thought that the complain was only on the name, but that you could re-use a decorator multiple times

**Reply by zzzeek** [reply]
I am going to fix this problem

**Comment by zzzeek** [comment]
The reason we have this thing where we have to keep renaming the property is due to https://github.com/python/typing/discussions/1102 .  However, making it harder is the behavior described at https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#defining-expression-behavior-distinct-from-attribute-behavior (yes in the 1.4 documentation) where we copy `@property` and have it return a **new object** each time.  This makes it very difficult to have a correctly typed example.

So I am introducing a modifier that will use the **old** behavior of hybrid_property, where it modifies in place, and document that.   

Here it is!   

```py
            class Interval(Base):
                # ...

                @hybrid_property
                def radius(self) -> float:
                    return abs(self.length) / 2

                @radius.inplace.setter
                def _radius_setter(self, value: float) -> None:
                    self.length = value * 2

                @radius.inplace.expression
                def _radius_expression(cls) -> ColumnElement[float]:
                    return type_coerce(func.abs(cls.length) / 2, Float)
```

**Reply by CaselIT** [reply]
ok, did not know that property behaved like this. but I guess I've never used property to add a deleter.
also with property you would call each method with the same name, so that problem does not exist.

The fact that type checkers have special cased only property is really sad :(

**Reply by michaelbukachi** [reply]
@zzzeek Is this already deployed? I saw the docs have already been updated. Also, what are the implications for querying? 
Do we use the hybrid_property method name:
`select(Interval.radius)`
or do we use the expression:
`select(Interval._radius_expression)`

**Reply by zzzeek** [reply]
i was just obsessing about people who would see the docs and think this was released yet.   nope, next hour hopefully

**Reply by michaelbukachi** [reply]
@zzzeek Using hybrid properties in `select` statements causes the following error:
```bash
...
TypeError: type BinaryExpression doesn't define __round__ method
```

Here's the sample snippet:
```python
import datetime

from sqlalchemy import create_engine, select, String, DateTime, ColumnElement, Float, \
    type_coerce, func
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import sessionmaker, DeclarativeBase, mapped_column, Mapped



class Base(DeclarativeBase):
    id = mapped_column(String, primary_key=True)
    created_at = mapped_column(DateTime, default=datetime.datetime.now)


class Item(Base):
    __tablename__ = 'items'

    quantity: Mapped[int]
    price: Mapped[float]

    @hybrid_property
    def total_price(self) -> float:
        return round(self.price * self.quantity, 2)

    @total_price.expression
    @classmethod
    def total_price_expr(cls) -> ColumnElement[float]:
        return type_coerce(func.round(cls.price * cls.quantity, 2), Float)




if __name__ == '__main__':
    engine = create_engine("sqlite:///")
    Base.metadata.create_all(engine)
    Session = sessionmaker(engine)

    with Session.begin() as db:
        stmt = select(Item.total_price).select_from(Item)
        db.execute(stmt)
```

**Reply by zzzeek** [reply]
your Item class defines the "expression" on an attribute called ``total_price_expr``, meaning if you were to select from `Item.total_price_expr` you'd get the complete hybrid you are looking for.  this is of course not the name you want so your options are to ensure you use the `total_price` name for the final method name, or to modify the original `total_price` hybrid without copying it using `total_price.inplace.expression`.  See https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#using-inplace-to-create-pep-484-compliant-hybrid-properties for background
